Services/RefundService.php

```diff
@@ -0,0 +1,281 @@
+use Doctrine\ORM\EntityManager;
+use Exception;
+use Mollie\Api\Resources\Order as MollieOrder;
+use Mollie\Api\Resources\Payment as MolliePayment;
+use Mollie\Api\Resources\OrderLine;
+use Mollie\Api\Resources\Refund;
+use MollieShopware\Models\OrderLines;
+use MollieShopware\Models\OrderLinesRepository;
+use MollieShopware\Models\Transaction;
+use MollieShopware\Traits\MollieApiClientTrait;
+use Shopware\Models\Order\Detail;
+use Shopware\Models\Order\Order;
+use Shopware\Models\Order\Status;
+
+class RefundService implements RefundInterface
+{
+    use MollieApiClientTrait;
+
+    /**
+     * @var EntityManager
+     */
+    private $modelManager;
+
+    public function __construct(EntityManager $modelManager)
+    {
+        $this->modelManager = $modelManager;
+    }
+
+    public function refundOrderAmount(Order $order, Transaction $transaction, float $customAmount = null)
+    {
+        $mollieClient = $this->getMollieApi($order->getShop()->getId());
+
+        if ($mollieClient === null) {
+            throw new \Exception('Something went wrong trying to get an API Client Instance. Please try again later.', 1);
+        }
+
+        if ($transaction->getMolliePaymentId() !== null) {
+            $molliePayment = $mollieClient->payments->get($transaction->getMolliePaymentId());
+
+            if (
+                (
+                    $customAmount === null ||
+                    $molliePayment->amount->value < $customAmount
+                ) && (
+                    $molliePayment->canBeRefunded() ||
+                    $molliePayment->canBePartiallyRefunded()
+                )
+            ) {
+                $this->refundPayment($order, $molliePayment);
+            }
+
+            if ($customAmount !== null && $molliePayment->canBePartiallyRefunded() && $molliePayment->getAmountRemaining() > $customAmount) {
+                $this->partialRefundPayment($order, $molliePayment, $customAmount);
+            }
+        }
+
+        if ($transaction->getMollieOrderId() !== null) {
+            /** @var MollieOrder $mollieOrder */
+            $mollieOrder = $mollieClient->orders->get($transaction->getMollieOrderId());
+
+            if ($customAmount === null || $mollieOrder->amountCaptured->value <= $customAmount) {
+                $this->refundOrder($order, $mollieOrder);
+            }
+
+            if ($customAmount !== null && $mollieOrder->amountCaptured->value > $customAmount) {
+                throw new Exception('partial refunds of orders is currently not implemented.');
+            }
+        }
+    }
+
+    /**
+     * Refund a Mollie order
+     *
+     * @param Order       $order
+     * @param MollieOrder $mollieOrder
+     *
+     * @return bool|Refund
+     * @throws \Exception
+     *
+     */
+    public function refundOrder(Order $order, MollieOrder $mollieOrder)
+    {
+        $apiClient = $this->getMollieApi($order->getShop()->getId());
+
+        if (empty($this->modelManager) || $apiClient === null) {
```

Comment:
if it works with fail-fast just remove it, or add an exception if the if is required, thx